### ISSUE TITLE
Replace COMPILE_TIME_ASSERT with static_assert

### ISSUE THREAD

**Issue by ckaran** [JuliaLang/libosxunwind#6]
Commit d470411fe7cae833bc93aaf90537c3646c2f605b fails on my machine with multiple errors along the following:

```
In file included from src/libuwind.cxx:33:
In file included from /Users/crossbar/Documents/Programming/Julia/libosxunwind/src/UnwindCursor.hpp:41:
/Users/crossbar/Documents/Programming/Julia/libosxunwind/src/DwarfInstructions.hpp:920:2: error: redeclaration of 'compile_time_assert_failed' with a different
      type: 'int [((int)CFI_Parser<A>::kMaxRegisterNumber > (int)DW_X86_64_RET_ADDR) ? 1 : -1]' vs 'int [1]'
        COMPILE_TIME_ASSERT( (int)CFI_Parser<A>::kMaxRegisterNumber > (int)DW_X86_64_RET_ADDR );
        ^
/Users/crossbar/Documents/Programming/Julia/libosxunwind/src/InternalMacros.h:51:14: note: expanded from macro 'COMPILE_TIME_ASSERT'
                extern int compile_time_assert_failed[ ( expr ) ? 1 : -1 ] __attribute__( ( unused ) );
                           ^
/Users/crossbar/Documents/Programming/Julia/libosxunwind/src/Registers.hpp:591:2: note: previous declaration is here
        COMPILE_TIME_ASSERT( sizeof(Registers_ppc) < sizeof(unw_context_t) );
        ^
/Users/crossbar/Documents/Programming/Julia/libosxunwind/src/InternalMacros.h:51:14: note: expanded from macro 'COMPILE_TIME_ASSERT'
                extern int compile_time_assert_failed[ ( expr ) ? 1 : -1 ] __attribute__( ( unused ) );
```

`COMPILE_TIME_ASSERT` is defined in `Julia/libosxunwind/src/InternalMacros.h` and is intended to be a compile-time assertion function.  This is precisely what `static_assert`, which is part of both the latest C and C++ standards, is intended for.  Replacing all instances of `COMPILE_TIME_ASSERT` with `static_assert` solves this problem.  The following patch does this:

```
From 51704b7d4fe633bc60bef174aef47044c12ce3bb Mon Sep 17 00:00:00 2001
From: Cem Karan <cem.f.karan.civ@mail.mil>
Date: Thu, 5 May 2016 09:25:37 -0400
Subject: [PATCH] Replaced COMPILE_TIME_ASSERT() with static_assert()

The former is a version of the latter, but the latter is part of the C and C++
standards, and so is supported on all compliant compilers.
---
 src/DwarfInstructions.hpp | 10 +++++++---
 src/Registers.hpp         | 10 +++++++---
 src/UnwindCursor.hpp      |  4 +++-
 3 files changed, 17 insertions(+), 7 deletions(-)

diff --git a/src/DwarfInstructions.hpp b/src/DwarfInstructions.hpp
index d063ebc..6784af8 100644
--- a/src/DwarfInstructions.hpp
+++ b/src/DwarfInstructions.hpp
@@ -29,6 +29,7 @@
 #ifndef __DWARF_INSTRUCTIONS_HPP__
 #define __DWARF_INSTRUCTIONS_HPP__
 
+#include <assert.h>
 #include <stdint.h>
 #include <stdio.h>
 #include <stdlib.h>
@@ -917,7 +918,8 @@ typename A::pint_t DwarfInstructions<A,R>::evaluateExpression(pint_t expression,
 template <typename A, typename R>
 int DwarfInstructions<A,R>::lastRestoreReg(const Registers_x86_64&) 
 {
-	COMPILE_TIME_ASSERT( (int)CFI_Parser<A>::kMaxRegisterNumber > (int)DW_X86_64_RET_ADDR );
+	static_assert( (int)CFI_Parser<A>::kMaxRegisterNumber > (int)DW_X86_64_RET_ADDR,
+	 "(int)CFI_Parser<A>::kMaxRegisterNumber MUST be > (int)DW_X86_64_RET_ADDR" );
 	return DW_X86_64_RET_ADDR; 
 }
 
@@ -1312,7 +1314,8 @@ compact_unwind_encoding_t DwarfInstructions<A,R>::createCompactEncodingFromProlo
 template <typename A, typename R>
 int DwarfInstructions<A,R>::lastRestoreReg(const Registers_x86&) 
 {
-	COMPILE_TIME_ASSERT( (int)CFI_Parser<A>::kMaxRegisterNumber > (int)DW_X86_RET_ADDR );
+	static_assert( (int)CFI_Parser<A>::kMaxRegisterNumber > (int)DW_X86_RET_ADDR,
+	"(int)CFI_Parser<A>::kMaxRegisterNumber MUST be > (int)DW_X86_RET_ADDR" );
 	return DW_X86_RET_ADDR; 
 }
 
@@ -1688,7 +1691,8 @@ compact_unwind_encoding_t DwarfInstructions<A,R>::createCompactEncodingFromProlo
 template <typename A, typename R>
 int DwarfInstructions<A,R>::lastRestoreReg(const Registers_ppc&) 
 {
-	COMPILE_TIME_ASSERT( (int)CFI_Parser<A>::kMaxRegisterNumber > (int)UNW_PPC_SPEFSCR );
+	static_assert( (int)CFI_Parser<A>::kMaxRegisterNumber > (int)UNW_PPC_SPEFSCR,
+	"(int)CFI_Parser<A>::kMaxRegisterNumber MUST be > (int)UNW_PPC_SPEFSCR" );
 	return UNW_PPC_SPEFSCR; 
 }
 
diff --git a/src/Registers.hpp b/src/Registers.hpp
index ad45ec8..c22d025 100644
--- a/src/Registers.hpp
+++ b/src/Registers.hpp
@@ -29,6 +29,7 @@
 #ifndef __REGISTERS_HPP__
 #define __REGISTERS_HPP__
 
+#include <assert.h>
 #include <stdint.h>
 #include <stdio.h>
 #include <stdlib.h>
@@ -106,7 +107,8 @@ private:
 
 inline Registers_x86::Registers_x86(const void* registers)
 {
-	COMPILE_TIME_ASSERT( sizeof(Registers_x86) < sizeof(unw_context_t) );
+	static_assert( sizeof(Registers_x86) < sizeof(unw_context_t),
+	"sizeof(Registers_x86) MUST be < sizeof(unw_context_t)" );
 	fRegisters = *((GPRs*)registers); 
 }
 
@@ -310,7 +312,8 @@ private:
 
 inline Registers_x86_64::Registers_x86_64(const void* registers)
 {
-	COMPILE_TIME_ASSERT( sizeof(Registers_x86_64) < sizeof(unw_context_t) );
+	static_assert( sizeof(Registers_x86_64) < sizeof(unw_context_t),
+	"sizeof(Registers_x86_64) MUST be < sizeof(unw_context_t)" );
 	fRegisters = *((GPRs*)registers); 
 }
 
@@ -588,7 +591,8 @@ private:
 
 inline Registers_ppc::Registers_ppc(const void* registers) 
 {
-	COMPILE_TIME_ASSERT( sizeof(Registers_ppc) < sizeof(unw_context_t) );
+	static_assert( sizeof(Registers_ppc) < sizeof(unw_context_t) ,
+	              "sizeof(Registers_ppc) MUST be < sizeof(unw_context_t)");
 	fRegisters = *((ppc_thread_state_t*)registers); 
 	fFloatRegisters = *((ppc_float_state_t*)((char*)registers+160));
 	memcpy(fVectorRegisters, ((char*)registers+424), sizeof(fVectorRegisters));
diff --git a/src/UnwindCursor.hpp b/src/UnwindCursor.hpp
index 637ef7b..e315ac3 100644
--- a/src/UnwindCursor.hpp
+++ b/src/UnwindCursor.hpp
@@ -29,6 +29,7 @@
 #ifndef __UNWINDCURSOR_HPP__
 #define __UNWINDCURSOR_HPP__
 
+#include <assert.h>
 #include <stdint.h>
 #include <stdio.h>
 #include <stdlib.h>
@@ -405,7 +406,8 @@ template <typename A, typename R>
 UnwindCursor<A,R>::UnwindCursor(unw_context_t* context, A& as)
   : fUnwindInfoMissing(false), fIsSignalFrame(false), fForceDwarf(false), fRegisters(context), fAddressSpace(as)
 {
-	COMPILE_TIME_ASSERT( sizeof(UnwindCursor<A,R>) < sizeof(unw_cursor_t) );
+	static_assert( sizeof(UnwindCursor<A,R>) < sizeof(unw_cursor_t),
+	"sizeof(UnwindCursor<A,R>) MUST be < sizeof(unw_cursor_t)" );
 
 	bzero(&fInfo, sizeof(fInfo));
 }
-- 
2.8.2

```

**Comment by tkelman** [comment]
Could you submit that as a pull request? Ping @Keno.

**Comment by ckaran** [comment]
Created, and pinged.  Let me know if you don't see it.

**Comment by tkelman** [comment]
closed by #7, thanks
